Add new audit event for DynamoDB protos

GavinFrazar opened the pull request:
Relevant issue: #15572

## What
This PR updates the events protos to include a new event: `AppSessionDynamoDBRequest`.

## Why
For some context:
* We plan to add DynamoDB database-access support eventually, but currently we only support dynamo via app-access.
* We want dynamo requests to show up in the audit log for database-access.
* We want dynamo request events to be consistently displayed in the web ui audit log, whether the requests were performed via app-access or database-access. Therefore we want dynamo requests via app-access to also be displayed in the web ui audit log.
* The already existing `AppSessionRequest` event is (intentionally) only emitted to a session file for upload, rather than being emitted directly to the audit log, and we want to keep that behavior. Therefore, we need a new event.

Additionally, we want to add some extra data to these dynamo request events:
1. ~~Operation: category of API operation. DynamoDB documents 4 kinds of operation: data, control, stream, transaction. This is a nice-to-have for better UX when I add web ui implementation.~~ edit: I'm removing this because dynamodb docs don't specify the category of all the API calls and there's 53 of them. Not worth the effort, will remove this and instead just add a `target` field which contains the `X-Amz-Target` header content. The format for that header is `<api>_<version>.<operation>`
2. Body: the json payload of each HTTP request. All dynamodb requests contain a JSON payload.
3. SessionChunkID: in database-access, we emit start/stop/query events with a session id that can be used to playback the db session with `tsh play --format json <session id>`. For app access, the ID needed for playback is the app session chunk id. Therefore, we include this field. Unlike `AppSessionRequest`, this event will be displayed in the web ui and for a consistent UX we need to provide a way for users to associate the event with its session chunk.

GavinFrazar committed:
protobuf update

GavinFrazar commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3693,6 +3695,81 @@ message ElasticsearchRequest {
+  // Operation represents the category of operation for an API being accessed in a given request.
+  OperationPlane Operation = 10 [(gogoproto.jsontag) = "operation"];
+
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 11 [
+    (gogoproto.nullable) = true,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
+  // App is a common application resource metadata.
+  AppMetadata App = 12 [
+    (gogoproto.nullable) = true,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
```
in particular, I have both metadata inside this proto message, but they are nullable `(gogoproto.nullable) = true` so we can choose which one to send (actually I could make them "one of" instead?)

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3693,6 +3695,81 @@ message ElasticsearchRequest {
+
+  // SessionChunkID is the ID of the session that was created for this 5 minute
+  // application log chunk.
+  string SessionChunkID = 3 [(gogoproto.jsontag) = "session_chunk_id"];
```
Do we still need session chunk ?

GavinFrazar replied:
Yes. In Database access we emit events to both the audit log and to a session for upload when the session ends. 

For example: 
<img width="1155" alt="image" src="https://user-images.githubusercontent.com/16846283/196264240-e0b700c1-36e5-4344-90a2-b6ac518cf5b9.png">

If I play back the session uploaded you can see all 3 events (start/query/end) are recorded within the session as well:

```
$ tsh play --format json 48d498b4-fc2e-405f-a345-f12386921d95 | jq                                  (base) 
{
  "cluster_name": "gavin-dev.cloud.gravitational.io",
  "code": "TDB00I",
  "db_protocol": "mysql",
  "db_service": "teleport-mysql",
  "db_uri": "teleport-mysql.mysql.database.azure.com:3306",
  "db_user": "teleport50",
  "ei": 0,
  "event": "db.session.start",
  "namespace": "default",
  "server_id": "08cf9ef5-e3be-49ff-9d05-6c8c0cb9e743",
  "sid": "48d498b4-fc2e-405f-a345-f12386921d95",
  "success": true,
  "time": "2022-10-17T00:40:40.084Z",
  "uid": "fb89307f-b033-443e-a446-8054e1bc8532",
  "user": "gavin.frazar@goteleport.com"
}
{
  "cluster_name": "gavin-dev.cloud.gravitational.io",
  "code": "TDB02I",
  "db_protocol": "mysql",
  "db_query": "select @@version_comment limit 1",
  "db_service": "teleport-mysql",
  "db_uri": "teleport-mysql.mysql.database.azure.com:3306",
  "db_user": "teleport50",
  "ei": 1,
  "event": "db.session.query",
  "sid": "48d498b4-fc2e-405f-a345-f12386921d95",
  "success": false,
  "time": "2022-10-17T00:40:40.146Z",
  "uid": "1a74c7b1-dc63-476a-b858-345821b4d1f1",
  "user": "gavin.frazar@goteleport.com"
}
{
  "cluster_name": "gavin-dev.cloud.gravitational.io",
  "code": "TDB01I",
  "db_protocol": "mysql",
  "db_service": "teleport-mysql",
  "db_uri": "teleport-mysql.mysql.database.azure.com:3306",
  "db_user": "teleport50",
  "ei": 2,
  "event": "db.session.end",
  "sid": "48d498b4-fc2e-405f-a345-f12386921d95",
  "time": "2022-10-17T00:40:41.181Z",
  "uid": "1ba7642c-f3d0-4809-83f4-6a99e019725f",
  "user": "gavin.frazar@goteleport.com"
}
```

GavinFrazar replied:
I can't think of a better way to define a "session" in app access. I realized it would be best to just keep the definition as a 5 minute TTL "chunk" of data.

One thing that you may have noticed is that `AppSessionRequest` does not include the session chunk id nor the session id.  This is because in the Web UI you do not see `AppSessionRequest` - you can only see them by playing back a session chunk.

For DynamoDB, since we want to see each request in the WebUI, I included the session chunk ID so that users can associate the log entry with the session chunk it is included in. This way, they can play back the associated chunk if they want to. Otherwise, determining which chunk a dynamo query came from would be awkward and difficult.

smallinsky replied:
Yes but in database access I don't think that  chunkID is present in `db.session.query` event in the current flow  for `tsh proxy aws` the [AppSessionRequest](https://github.com/gravitational/teleport/blob/a75fcc21d87c6d06ba7d8b358150e5a78d55e0f3/lib/srv/app/aws/handler.go#L165)  doesn't contain chunkID.

GavinFrazar replied:
Reposting here since we discussed this in slack: db start/end/query events all contain the session id, which is used to play back a db session. For app access, it's the session chunk id that is used for session playback. Therefore we are keeping the chunk id in the event for better UX.

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3693,6 +3695,81 @@ message ElasticsearchRequest {
+  string SessionChunkID = 3 [(gogoproto.jsontag) = "session_chunk_id"];
+
+  // AWS contains extra AWS metadata of the request.
+  AWSRequestMetadata AWS = 4 [
```
AWSRequestMetadata contains the servicename and the region. In case of DynamoDB access the service is constant.

GavinFrazar replied:
hmm but region and host are not constant...

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3693,6 +3695,81 @@ message ElasticsearchRequest {
+// DynamoDBRequest is emitted when a user executes a DynamoDB request.
+// We support DynamoDB access via app access and database access,
+// therefore this event includes either DatabaseMetadata or AppMetadata.
+message DynamoDBRequest {
```
We need to introduce event name and event code for `DynamoDBRequest` audit event.

GavinFrazar replied:
Yep, will be added in a later PR, this one is just to get the protos definition nailed down

GavinFrazar committed:
Update proto to use dynamodb request event specific to app-access

We will include a similar event for dynamodb via database-access.
We split the events so that app and database access events are not coupled.
This way we do not have to include optional database/app metadata in one event too.

GavinFrazar committed:
Update protos

GavinFrazar committed:
Update oneof

GavinFrazar committed:
Move AppMetaData up with the other metadata and add a 'target' field

GavinFrazar committed:
Remove operation plane

GavinFrazar commented:
@smallinsky this is ready, please take another look when you can

Tener commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3693,6 +3695,56 @@ message ElasticsearchRequest {
+  // The Struct type is a wrapper around protobuf/types.Struct and is used to marshal the JSON body correctly.
+  google.protobuf.Struct Body = 11 [
+    (gogoproto.jsontag) = "body,omitempty",
+    (gogoproto.casttype) = "Struct"
```
Custom proto serialization via `casttype` can fail, it might be good idea to test this round trips etc.

GavinFrazar replied:
This is the same way we serialize IdentityAttributes in UserLogin message, which is where I got the idea to do this. I have manually verified that it serializes/deserializes properly.

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/improve_dynamodb_audit_events_proto

GavinFrazar commented:
@zmb3 @smallinsky need approval from one of you since other approvals were both group 2

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/improve_dynamodb_audit_events_proto

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/improve_dynamodb_audit_events_proto

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/improve_dynamodb_audit_events_proto
